### PR TITLE
feat: support reconstructing chain of certs

### LINKED ISSUE #156: PEM certificate extracted from PFX is in the wrong order
**What steps did you take and what happened:**
Added a PFX certificate which contains a certificate chain with server, intermediate, and root certificate. The PEM file had the structure of:

PRIVATE KEY
ROOT
INTERMEDIATE
SERVER

**What did you expect to happen:**

According to https://tools.ietf.org/html/rfc4346#section-7.4.2 the certificate order should be

SERVER
INTERMEDIATE
ROOT
KEY

If you run openssl x509 -in downloaded.pem -noout -text on the downloaded certificate you'll see the root cert.

**Anything else you would like to add:**
We should also have the option to download the PFX certificate as is.

**Which access mode did you use to access the Azure Key Vault instance:**
Pod Identity

**Environment:**

- Secrets Store CSI Driver version: (use the image tag):
- Azure Key Vault provider version: 0.0.7:
- Kubernetes version: 1.16.10
- Cluster type: AKS

### PR DESCRIPTION
**Pull request by aramase**
**Reason for Change**:

- Adds an optional feature `ConstructPEMChain` to reconstruct the chain in the AKV provider.

**Requirements**

- [x] squashed commits
- [x] included documentation
- [x] added unit tests and e2e tests (if applicable).

**Issue Fixed**:

fixes #156 

**Please answer the following questions with yes/no**:

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?

**Special Notes for Reviewers**:
Reference docs:
https://social.technet.microsoft.com/wiki/contents/articles/3147.pki-certificate-chaining-engine-cce.aspx#Building_the_Certificate_Chain

### CODE REVIEW

**Comment by aramase** [comment]
cc @RandalliLama

File: pkg/provider/provider.go [hunk 1]
```diff
@@ -621,3 +636,83 @@ func validateFileName(fileName string) error {
...
+	for {
+		// decode pem to der first
+		block, rest := pem.Decode(data)
+		data = rest
+
+		if block == nil {
+			break
+		}
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
```
Comment (ritazh) [inline pkg/provider/provider.go:660]: maybe also explicitly check for nil cert

File: pkg/provider/provider.go [hunk 2]
```diff
@@ -621,3 +636,83 @@ func validateFileName(fileName string) error {
...
+		return nil, fmt.Errorf("no leaf found")
+	}
+
+	// iterate through the directed list and append the nodes to new cert chain
+	for leaf != nil {
+		newCertChain = append(newCertChain, leaf.cert)
+		leaf = leaf.parent
+	}
+
+	for _, cert := range newCertChain {
```
Comment (ritazh) [inline pkg/provider/provider.go:710]: Do we need to validate that all three parts are included in newCertChain?

Comment (RandalliLama) [reply pkg/provider/provider.go:710]: @ritazh Do you mean to validate that there is a leaf, an intermediate, and a CA.  The code should not do that as not all PEM files will have all three and they are not required to.  They may only contain a leaf certificate.

File: pkg/provider/provider.go [hunk 3]
```diff
@@ -621,3 +636,83 @@ func validateFileName(fileName string) error {
...
+			break
+		}
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return pemData, err
+		}
+		nodes = append(nodes, &node{
+			cert:     cert,
+			parent:   nil,
+			isParent: false,
```
Comment (RandalliLama) [inline pkg/provider/provider.go:666]: I'm curious why you need both 'parent' and 'isParent'.  Isn't that redundant.  Can you just check parent for nil?  Or is this a Go idiom?

**Review by RandalliLama — APPROVED** [review]
Left a couple a minor comments.

File: pkg/provider/provider.go (hunk 2 above)
Comment (ritazh) [reply pkg/provider/provider.go:710]: > The code should not do that as not all PEM files will have all three and they are not required to.

Gotcha!

File: pkg/provider/provider.go (hunk 3 above)
Comment (aramase) [reply pkg/provider/provider.go:666]: The idea was to do the leaf cert lookup O(n) using `isParent`. If `isParent` is false, then that's the leaf node and then we just construct the chain until we hit the `nil` node.

File: pkg/provider/provider.go (hunk 1 above)
Comment (aramase) [reply pkg/provider/provider.go:660]: Added the nil cert check.

**Comment by thebenwaters** [comment]
Hey guys, while this is helpful for a PEM formatted file, the built in go pkcs12 library does not support multiple certs inside pkcs12 file. 

This library

https://pkg.go.dev/software.sslmate.com/src/go-pkcs12 is a fork licensed to bsd-3

And supports multiple certificates inside of a pkcs12 file

I'm going to open a PR with attempts to decipher format and parse it out. I'm contacting you all internally (MSFT)

**Comment by thebenwaters** [comment]
After re-review @aramase basically implements the parsing/reconstruct algo work that is done inside of this package.

**Review by ritazh — APPROVED** [review]

**Pull request merged**